state-sync/aptos-data-client/src/aptosnet/mod.rs

```diff
@@ -146,46 +156,70 @@ impl AptosNetDataClient {
             })
     }
 
-    /// Fetches the next group of peers to poll. The group will contain: (i) any (new) peers that
-    /// have connected since the last time this method was called (i.e., the peers that have not
-    /// been polled yet); (ii) at most one prioritized peer (e.g., those that are upstream); and
-    /// (iii) at most one non-prioritized peer (i.e., those that are downstream).
-    fn fetch_peers_to_poll(&self) -> Result<Vec<PeerNetworkId>, Error> {
-        let mut peers_to_poll = vec![];
+    /// Fetches the next prioritized peer to poll
+    fn fetch_prioritized_peer_to_poll(&self) -> Result<Option<PeerNetworkId>, Error> {
+        self.select_peer_to_poll(true)
+    }
 
-        // Fetch the last polled high-priority peer
-        if let Some(peer) = self.peer_states.write().oldest_polled_priority_peer() {
-            peers_to_poll.push(peer);
-        }
+    /// Fetches the next regular peer to poll
+    fn fetch_regular_peer_to_poll(&self) -> Result<Option<PeerNetworkId>, Error> {
+        self.select_peer_to_poll(false)
+    }
 
-        // Fetch all new peers (i.e., those not yet polled)
-        for peer in self.get_all_connected_peers()? {
-            if !self.peer_states.read().already_polled_peer(&peer) {
-                peers_to_poll.push(peer);
-            }
-        }
+    /// Randomly selects a peer to poll that does not have an in-flight request. Peer
+    /// filtering is done based on `is_priority_peer`.
+    fn select_peer_to_poll(&self, is_priority_peer: bool) -> Result<Option<PeerNetworkId>, Error> {
+        // Fetch the number of in-flight polls
+        let num_in_flight_polls = if is_priority_peer {
```

Comment:
nit: this doesn't look so clean - we seem to be checking for `is_priority_peer` for almost every line of the function. Maybe the previous implementation with two separate functions was a cleaner approach? I would leave it to you to make a call. Sorry about the back and forth.

Reply:
Agreed -- I rolled back to the previous version 😄